controllers/addon_controller_test.go

```diff
@@ -27,114 +25,102 @@ var (
 
 const timeout = time.Second * 5
 
+var instance *v1alpha1.Addon
+var updated *v1alpha1.Addon
+var fetchedAddon *v1alpha1.Addon
+var addonController *Controller
+
 var _ = Describe("AddonController", func() {
 
 	Describe("Addon CR can be reconciled", func() {
-		var instance *v1alpha1.Addon
-		var wfv1 = &unstructured.Unstructured{}
-		wfv1.SetGroupVersionKind(schema.GroupVersionKind{
-			Kind:    "Workflow",
-			Group:   "argoproj.io",
-			Version: "v1alpha1",
-		})
-
 		It("instance should be parsable", func() {
-			addonYaml, err := ioutil.ReadFile("../docs/examples/clusterautoscaler.yaml")
+			addonYaml, err := ioutil.ReadFile("./tests/clusterautoscaler.yaml")
 			Expect(err).ToNot(HaveOccurred())
 
 			instance, err = parseAddonYaml(addonYaml)
 			Expect(err).ToNot(HaveOccurred())
 			Expect(instance).To(BeAssignableToTypeOf(&v1alpha1.Addon{}))
 			Expect(instance.GetName()).To(Equal(addonName))
+
 		})
 
 		It("instance should be reconciled", func() {
 			instance.SetNamespace(addonNamespace)
-			err := k8sClient.Create(context.TODO(), instance)
-			if apierrors.IsInvalid(err) {
-				Fail(fmt.Sprintf("failed to create object, got an invalid object error. %v", err))
-			}
-			Expect(err).NotTo(HaveOccurred())
-
-			Eventually(func() error {
-				if err := k8sClient.Get(context.TODO(), addonKey, instance); err != nil {
-					return err
-				}
+			addonController = newController(instance)
 
-				if len(instance.ObjectMeta.Finalizers) > 0 {
-					return nil
-				}
-				return fmt.Errorf("addon is not valid")
-			}, timeout).Should(Succeed())
+			processed := addonController.processNextItem(ctx)
+			Expect(processed).To(BeTrue())
+			var objects []runtime.Object
+			wfcli := wfclientsetfake.NewSimpleClientset(objects...)
```

Comment:
There's no reason to use a fake here, I don't see why we even had to rewrite this test logic, it should've worked given the previous implementation b/c it should simply test the reconciliation states.

Reply:
As you can see, all the tests function retained exactly the same (verify how addon works) --- reconcile an Addon, verify its checksum, TTL; modify the spec and trigger addon-reinstall; dependency package checking and process. 

The wfcli provides a way to access the Addon API, which is needed by the controller frameworkl